diff --git a/pkgs/os-specific/linux/musl/default.nix b/pkgs/os-specific/linux/musl/default.nix
index bef1fa6e33a..409cfb64f33 100644
--- a/pkgs/os-specific/linux/musl/default.nix
+++ b/pkgs/os-specific/linux/musl/default.nix
@@ -114,6 +114,15 @@ stdenv.mkDerivation rec {
     ${lib.optionalString (stdenv.targetPlatform.libc == "musl" && stdenv.targetPlatform.isx86_32)
       "install -D libssp_nonshared.a $out/lib/libssp_nonshared.a"
     }
+
+    # For some reason, on M1 Mac, there is exists $out/lib/ld-musl-x86_64.so.1 symlink,
+    # which has incorrect permissions (isn't readable). We can't change permissions
+    # on a symlink using busybox's chmod: it lacks -h option.
+    # Try to change permissions on source libc.so and recreate the link, it would change permissions
+    # on the resulting symlink
+    rm -f $out/lib/ld-musl-${stdenv.hostPlatform.uname.processor}.so.1
+    chmod 0555 $out/lib/libc.so
+    ln -rs $out/lib/libc.so $out/lib/ld-musl-${stdenv.hostPlatform.uname.processor}.so.1

     # Create 'ldd' symlink, builtin
     ln -rs $out/lib/libc.so $out/bin/ldd
