diff --git a/pkgs/development/compilers/gcc/11/default.nix b/pkgs/development/compilers/gcc/11/default.nix
index 3a61c5820b1..b4638f6b18b 100644
--- a/pkgs/development/compilers/gcc/11/default.nix
+++ b/pkgs/development/compilers/gcc/11/default.nix
@@ -46,13 +46,7 @@ with lib;
 with builtins;

 let majorVersion = "11";
-    # The patch below for aarch64-darwin does not apply to 11.3.0 and an
-    # updated version is not available. Keep aarch64-darwin on 11.2.0 so the
-    # large body of packages which depend on gfortran are still functional
-    # until GCC 12 is the default.
-    # On x86_64-darwin, building libgcc suffers from some different issues with 11.3.0.
-    version = if stdenv.isDarwin then
-      "${majorVersion}.2.0" else "${majorVersion}.3.0";
+    version = "${majorVersion}.3.0";

     inherit (stdenv) buildPlatform hostPlatform targetPlatform;

@@ -69,10 +63,14 @@ let majorVersion = "11";
       ++ optional langFortran ../gfortran-driving.patch
       ++ optional (targetPlatform.libc == "musl" && targetPlatform.isPower) ../ppc-musl.patch

-      ++ optional (stdenv.isDarwin && stdenv.isAarch64) (fetchpatch {
-        url = "https://github.com/fxcoudert/gcc/compare/releases/gcc-11.2.0...gcc-11.2.0-arm-20211201.diff";
-        sha256 = "sha256-z62s/cXuH9Kgq/oD/OiiZ8LWnX1xl1D43sONnwaEW1w=";
-      })
+      ++ optionals stdenv.isDarwin [
+        (fetchpatch {
+          # Corresponds to https://github.com/iains/gcc-11-branch/releases/tag/gcc-11.3-darwin-r2
+          url = "https://github.com/Homebrew/formula-patches/raw/8184603db1db7da12d0fbfa47e1329ba0e21558e/gcc/gcc-11.3.0-arm.diff";
+          sha256 = "sha256-sEClJ7BzXJccLm+V6LF9Ioe4o/aV3iJ4akvOVoPG5pA=";
+        })
+        ./libgcc-darwin-libemutls-no-shared.patch
+      ]

       # Obtain latest patch with ../update-mcfgthread-patches.sh
       ++ optional (!crossStageStatic && targetPlatform.isMinGW) ./Added-mcf-thread-model-support-from-mcfgthread.patch;
@@ -92,9 +90,7 @@ stdenv.mkDerivation ({

   src = fetchurl {
     url = "mirror://gcc/releases/gcc-${version}/gcc-${version}.tar.xz";
-    sha256 = if stdenv.isDarwin
-      then "sha256-0I7cU2tUw3KhAQ/2YZ3SdMDxYDqkkhK6IPeqLNo2+os="
-      else "sha256-tHzygYaR9bHiHfK7OMeV+sLPvWQO3i0KXhyJ4zijrDk=";
+    sha256 = "sha256-tHzygYaR9bHiHfK7OMeV+sLPvWQO3i0KXhyJ4zijrDk=";
   };

   inherit patches;
diff --git a/pkgs/development/compilers/gcc/11/libgcc-darwin-libemutls-no-shared.patch b/pkgs/development/compilers/gcc/11/libgcc-darwin-libemutls-no-shared.patch
index e69de29bb2d..78e26ef5299 100644
--- a/pkgs/development/compilers/gcc/11/libgcc-darwin-libemutls-no-shared.patch
+++ b/pkgs/development/compilers/gcc/11/libgcc-darwin-libemutls-no-shared.patch
@@ -0,0 +1,28 @@
+From dc655807b03680ff301bf3df53daa973e63af354 Mon Sep 17 00:00:00 2001
+From: Zhaofeng Li <hello@zhaofeng.li>
+Date: Tue, 18 Oct 2022 03:20:54 -0600
+Subject: [PATCH] Darwin, libgcc: Only build libemutls_w.a with shared
+
+---
+ libgcc/config.host | 5 ++++-
+ 1 file changed, 4 insertions(+), 1 deletion(-)
+
+diff --git a/libgcc/config.host b/libgcc/config.host
+index 771a42cfb89..25812d910ef 100644
+--- a/libgcc/config.host
++++ b/libgcc/config.host
+@@ -264,7 +264,10 @@ case ${host} in
+       echo "Warning: libgcc configured to support macOS 10.5" 1>&2
+       ;;
+   esac
+-  extra_parts="crt3.o crttms.o crttme.o libemutls_w.a"
++  extra_parts="crt3.o crttms.o crttme.o"
++  if test x$enable_shared = xyes; then
++    extra_parts="$extra_parts libemutls_w.a"
++  fi
+   ;;
+ *-*-dragonfly*)
+   tmake_file="$tmake_file t-crtstuff-pic t-libgcc-pic t-eh-dw2-dip"
+--
+2.37.3
+
diff --git a/pkgs/development/libraries/gcc/libgcc/default.nix b/pkgs/development/libraries/gcc/libgcc/default.nix
index 094bb7e7a1d..23207ec0124 100644
--- a/pkgs/development/libraries/gcc/libgcc/default.nix
+++ b/pkgs/development/libraries/gcc/libgcc/default.nix
@@ -5,7 +5,7 @@

 stdenvNoLibs.mkDerivation rec {
   pname = "libgcc";
-  inherit (gcc.cc) src version;
+  inherit (gcc.cc) src version patches;

   outputs = [ "out" "dev" ];

diff --git a/pkgs/os-specific/linux/musl/default.nix b/pkgs/os-specific/linux/musl/default.nix
index bef1fa6e33a..409cfb64f33 100644
--- a/pkgs/os-specific/linux/musl/default.nix
+++ b/pkgs/os-specific/linux/musl/default.nix
@@ -114,6 +114,15 @@ stdenv.mkDerivation rec {
     ${lib.optionalString (stdenv.targetPlatform.libc == "musl" && stdenv.targetPlatform.isx86_32)
       "install -D libssp_nonshared.a $out/lib/libssp_nonshared.a"
     }
+
+    # For some reason, on M1 Mac, there is exists $out/lib/ld-musl-x86_64.so.1 symlink,
+    # which has incorrect permissions (isn't readable). We can't change permissions
+    # on a symlink using busybox's chmod: it lacks -h option.
+    # Try to change permissions on source libc.so and recreate the link, it would change permissions
+    # on the resulting symlink
+    rm -f $out/lib/ld-musl-${arch}.so.1
+    chmod 0555 $out/lib/libc.so
+    ln -rs $out/lib/libc.so $out/lib/ld-musl-${arch}.so.1

     # Create 'ldd' symlink, builtin
     ln -rs $out/lib/libc.so $out/bin/ldd

